plugin/path_reserves.go

```diff
@@ -0,0 +1,312 @@
+			},
+		},
+		HelpSynopsis:    pathListReservesHelpSyn,
+		HelpDescription: pathListReservesHelpDesc,
+	}
+}
+
+func (b *backend) reserveListOperation(ctx context.Context, req *logical.Request, _ *framework.FieldData) (*logical.Response, error) {
+	keys, err := req.Storage.List(ctx, libraryPrefix)
+	if err != nil {
+		return nil, err
+	}
+	return logical.ListResponse(keys), nil
+}
+
+func (b *backend) pathReserves() *framework.Path {
+	return &framework.Path{
+		Pattern: libraryPrefix + framework.GenericNameRegex("name"),
+		Fields: map[string]*framework.FieldSchema{
+			"name": {
+				Type:        framework.TypeLowerCaseString,
+				Description: "Name of the reserve",
+				Required:    true,
+			},
+			"service_account_names": {
+				Type:        framework.TypeCommaStringSlice,
+				Description: "The username/logon name for the service accounts with which this reserve will be associated.",
+			},
+			"lending_period": {
+				Type:        framework.TypeDurationSecond,
+				Description: "In seconds, the default length of time before check-outs will expire.",
+				Default:     24 * 60 * 60, // 24 hours
+			},
+		},
+		Operations: map[logical.Operation]framework.OperationHandler{
+			logical.CreateOperation: &framework.PathOperation{
+				Callback: b.operationReserveCreate,
+				Summary:  "Create a library reserve.",
+			},
+			logical.UpdateOperation: &framework.PathOperation{
+				Callback: b.operationReserveUpdate,
+				Summary:  "Update a library reserve.",
+			},
+			logical.ReadOperation: &framework.PathOperation{
+				Callback: b.operationReserveRead,
+				Summary:  "Read a library reserve.",
+			},
+			logical.DeleteOperation: &framework.PathOperation{
+				Callback: b.operationReserveDelete,
+				Summary:  "Delete a library reserve.",
+			},
+		},
+		ExistenceCheck:  b.operationReserveExistenceCheck,
+		HelpSynopsis:    reserveHelpSynopsis,
+		HelpDescription: reserveHelpDescription,
+	}
+}
+
+func (b *backend) operationReserveExistenceCheck(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (bool, error) {
+	reserve, err := readReserve(ctx, req.Storage, fieldData.Get("name").(string))
+	if err != nil {
+		return false, err
+	}
+	return reserve != nil, nil
+}
+
+func (b *backend) operationReserveCreate(ctx context.Context, req *logical.Request, fieldData *framework.FieldData) (*logical.Response, error) {
+	reserveName := fieldData.Get("name").(string)
+	serviceAccountNames := fieldData.Get("service_account_names").([]string)
+	lendingPeriodRaw, lendingPeriodSent := fieldData.GetOk("lending_period")
+	if !lendingPeriodSent {
+		lendingPeriodRaw = fieldData.Schema["lending_period"].Default
+	}
+	lendingPeriod := time.Duration(lendingPeriodRaw.(int)) * time.Second
+
+	if len(serviceAccountNames) == 0 {
+		return logical.ErrorResponse(`"service_account_names" must be provided`), nil
+	}
+	if err := ensureNotInAnotherReserve(ctx, req.Storage, serviceAccountNames); err != nil {
+		return nil, err
```

Comment:
Hi! Yes, I agree, it's racy. I was actually planning to address raciness in a different PR - basically, one that's after all the endpoints have been written. I think the one finalizing the `CheckOutHandler` model would be the best one to go with. I'll add a TODO in the code so there's no way this feature can be merged into master without addressing raciness, because it'll be super evident in the diff that it still needs to be addressed.

Regarding 4xx or 5xx here, it could go either way because the _other_ error that could be returned would be if the storage layer went down during a Get/Put operation. Since the type of error is unclear here, I return it as a 5xx. I suppose I could return a strongly typed error indicating it's in another reserve, and if it's that error, return a 4xx.

I think, though, that I'll instead strike this function and use the `CheckOutHandler` `Status` call to check if a service account already exists, and change the contract there.